[ITensors] Support for MPO*MPO contraction by densitymatrix method

shinaoka opened the pull request:
# Description

Based on discussions with Miles.

This small patch allows contracting an MPO and an MPS with a dangling bond at each site. This is intended to be a minimal and tentative patch that will eventually be replaced by ITensorNetworks.jl.

<details><summary>Minimal demonstration of previous behavior</summary><p>

```julia
using ITensors

R = 3
sites = [Index(2, "Qubit,n=$n") for n in 1:R]
a = replaceprime(randomMPO(sites), 0 => 1, 1 => 2)
b = randomMPO(sites)
ab_ref = contract(a, b; alg="naive")
ab = contract(a, b; alg="densitymatrix")

ERROR: MethodError: no method matching contract(::NDTensors.Algorithm{:densitymatrix}, ::MPO, ::MPO)

Closest candidates are:
  contract(::NDTensors.Algorithm{:naive}, ::MPO, ::MPO; kwargs...)
   @ ITensors ~/.julia/packages/ITensors/HjjU3/src/mps/mpo.jl:801
  contract(::NDTensors.Algorithm{:densitymatrix}, ::MPO, ::MPS; kwargs...)
   @ ITensors ~/.julia/packages/ITensors/HjjU3/src/mps/mpo.jl:678
  contract(::NDTensors.Algorithm{:zipup}, ::MPO, ::MPO; kwargs...)
   @ ITensors ~/.julia/packages/ITensors/HjjU3/src/mps/mpo.jl:805
  ...

Stacktrace:
 [1] contract(A::MPO, B::MPO; alg::String, kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
   @ ITensors ~/.julia/packages/ITensors/HjjU3/src/mps/mpo.jl:798
 [2] top-level scope
   @ REPL[7]:1
```
</p></details>

<details><summary>Minimal demonstration of new behavior</summary><p>

```julia
using ITensors

R = 3
sites = [Index(2, "Qubit,n=$n") for n in 1:R]
a = replaceprime(randomMPO(sites), 0 => 1, 1 => 2)
b = randomMPO(sites)
ab_ref = contract(a, b; alg="naive")
ab = contract(a, b; alg="densitymatrix")
ab_ref ≈ ab
```
</p></details>

# How Has This Been Tested?

A test will be added in `test_mpo.jl` that compares the results of contractions of an MPO and an MPS with a dangling bond.

# Checklist:

- [x] My code follows the style guidelines of this project. Please run `using JuliaFormatter; format(".")` in the base directory of the repository (`~/.julia/dev/ITensors`) to format your code according to our style guidelines.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that verify the behavior of the changes I made.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] Any dependent changes have been merged and published in downstream modules.

mtfishman commented:
Thanks @shinaoka, this will be useful to have.

I would instead vote for generalizing the function signature of:
```julia
function contract(::Algorithm"densitymatrix", A::MPO, ψ::MPS; kwargs...)
  # [...]
end
```
to:
```julia
function contract(::Algorithm"densitymatrix", A::AbstractMPS, B::AbstractMPS; kwargs...)
  # [...]
end
```
and then returning an object that is `typeof(B)`.

I'm also not such a fan of:
```julia
  function _getsimψ_c(i)
    t = copy(ψ_c[i])
    replaceinds!(t, dangling_inds_ψ[i], sim_dangling_inds_ψ[i])
    return t
  end
```
Why not just use:
```julia
replaceinds(ψ_c[i], dangling_inds_ψ[i], sim_dangling_inds_ψ[i])
```
directly?

shinaoka commented:
Thank you for the suggestion. I have addressed the issues.
I am a bit afraid of "Merging #1123 (3a4dcfd) into main (7120090) will decrease coverage by 29.29%.".
What do you think?

emstoudenmire commented on src/mps/mpo.jl:
```diff
@@ -1051,3 +1071,10 @@ function HDF5.read(parent::Union{HDF5.File,HDF5.Group}, name::AbstractString, ::
 end
+
+function contract(::Algorithm"densitymatrix", M1::MPO, M2::MPO; kwargs...)::MPO
+  t1 = MPO([M1[v] for v in eachindex(M1)])
```
I wanted to understand what this line is doing. Since `M1` is guarateed to be an MPO from the function signature, is `t1` needed?

shinaoka replied:
Thank you for pointing out this. This function is no more required as the new `contract` function accepts a pair of two AbstractMPS objects.

shinaoka commented on test/ITensorLegacyMPS/base/test_mpo.jl:
```diff
@@ -814,4 +814,14 @@ end
   end
 end
 
+@testset "MPO*MPO contraction (densitymatrix)" begin
```
I will move this test into inside the testset "MPO Basics".

shinaoka commented:
I have addressed the issues:

* The testset has been moved into inside the testset "MPO Basics".
* "function contract(::Algorithm"densitymatrix", M1::MPO, M2::MPO; kwargs...)::MPO" has been removed.

https://github.com/shinaoka/ITensors.jl/commit/6098a1067eb02e3bb6f9fce45a572547a83acd65

emstoudenmire commented:
[test ITensors mps]  (this comment ought to run MPS and MPO tests also)

emstoudenmire commented:
Looks good to me @shinaoka. Let's see how the tests do and if Matt has more comments.

shinaoka commented:
One test on automatic differentiation failed. I got precisely the same error on my Macbook Pro with the latest commit in the main branch. The error seems to be independent of this PR.
